phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -572,40 +577,55 @@ private void prepareToRebuildIndexes(Connection conn, String dataTableFullName)
             }
 
             if (hasViewIndex) {
-                ResultSet
-                        rs =
-                        conn.createStatement().executeQuery(
-                                "SELECT DISTINCT TABLE_NAME, TENANT_ID FROM "
-                                        + "SYSTEM.CATALOG WHERE COLUMN_FAMILY = \'"
-                                        + viewIndexPhysicalName
-                                        + "\' AND TABLE_TYPE = \'i\' AND " + "LINK_TYPE = "
-                                        + PTable.LinkType.PHYSICAL_TABLE.getSerializedValue());
+                String viewSql = "SELECT DISTINCT TABLE_NAME, TENANT_ID FROM "
+                        + "SYSTEM.CATALOG "
+                        + "WHERE COLUMN_FAMILY = \'" + dataTableFullName + "\' "
+                        + "AND LINK_TYPE = "
+                        + PTable.LinkType.PHYSICAL_TABLE.getSerializedValue();
+
+                ResultSet rs =
+                        conn.createStatement().executeQuery(viewSql);
+
                 while (rs.next()) {
-                    String viewIndexName = rs.getString(1);
+                    String viewName = rs.getString(1);
                     String tenantId = rs.getString(2);
-                    ResultSet
-                            innerRS =
-                            conn.createStatement().executeQuery(
-                                    "SELECT DISTINCT TABLE_NAME FROM "
-                                            + "SYSTEM.CATALOG WHERE COLUMN_FAMILY = \'"
-                                            + viewIndexName
-                                            + "\' AND TABLE_TYPE = \'i\' AND " + "LINK_TYPE = "
-                                            + PTable.LinkType.INDEX_TABLE.getSerializedValue());
-                    innerRS.next();
-                    String viewName = innerRS.getString(1);
-                    IndexInfo indexInfo = new IndexInfo(schemaName, viewName, tenantId == null ? GLOBAL_INDEX_ID: tenantId,
-                            viewIndexName);
-                    rebuildIndexes.put(viewIndexName, indexInfo);
+
+                    findViewIndexes(conn, schemaName, viewName, tenantId, rebuildIndexes);
                 }
             }
             //for rebuilding indexes in case of upgrade.
             rebuildMap.put(dataTableFullName, rebuildIndexes);
+
+            String json = gson.toJson(rebuildMap);
+            LOGGER.info("Index rebuild map " + json);
         } catch (SQLException e) {
-            LOGGER.severe("Failed to prepare the map for index rebuilds "+e);
+            LOGGER.severe("Failed to prepare the map for index rebuilds " + e);
             throw new RuntimeException("Failed to prepare the map for index rebuilds");
         }
     }
 
+    private void findViewIndexes(Connection conn, String schemaName, String viewName,
+            String tenantId, HashMap<String, IndexInfo> rebuildIndexes) throws SQLException {
```

Comment:
One more request. Return list<string> with view index names from this helper func. Create IndexInfo's and populate rebuildIndexes in the caller.

Reply:
Or if you think IndexInfo struct would be generally reusable, you can perhaps create the IndexInfo's in the helper and just populate rebuildIndexes in the caller.

Reply:
This would also mean changing the rebuild indexes to have an array list of indexeInfo.

Reply:
Ok. Then just return list of viewIndexName's and construct IndexInfo here and populate rebuildIndexes here. It shouldn't require changing the rebuildIndexes structure. You will just need to iterate over the returned list of viewIndexes.

Reply:
I see what you mean, thanks! :D

Reply:
Likely, we raced in commenting. But, I also thought the same and updated the PR :)